### DISCUSSION TITLE
How to turn signal back into owned value?

### DISCUSSION THREAD

**Discussion by stefnotch** [leptos-rs/leptos#3066] [Q&A] [answered]
I've been going through the Leptos counter examples, both v0.6 and master.
There's an `add_counter` function, akin to `Vec::push`. Then there's a `clear_counters` function, which resembles `Vec::clear`. So I figured I'd take all of that, and wrap it in a custom "signal vector" struct.
```rust
pub struct SignalVec<T>
where
    T: 'static,
{
    inner: RwSignal<Vec<RwSignal<T>>>,
}
```

Implementing `push` is straightfoward. Implementing `get` is also doable, even if it triggers reactivity more often than necessary.

But how would I implement `pop`?

```rust
pub fn push(&self, value: T) {
    self.inner.update(|inner| {
        inner.push(create_rw_signal(value));
    });
}

pub fn pop(&self) -> Option<T> {
    self.inner.update(|inner| {
        let signal = inner.pop();
        // And now what?
        // signal.dispose_and_take_underlying_value()
    });
}
```

Is this simply a missing API, or is there a deeper reason as to why this would not work?

**Comment by gbj** [comment, marked as answer]
`.try_update` returns a value, so I guess you could `.get()` the signal. No, there's not a "consume the signal and give me the value" function, because of the arena-based (rather than reference-counted) storage -- there's no guarantee that you don't have other copies of the inner `RwSignal` lying around so it's not safe to consume the value unless it's disposed. (Whether disposing of a signal should give you its inner value is an interesting question, and not the bad idea, but the first time it's been suggested.)

**Reply by stefnotch** [reply]
Thank you, that makes sense. So my options are either cloning the inner value, or putting it in an Option and taking that. 

Should I open an issue regarding getting back the inner value when disposing of a signal? Mainly to track the idea, if it makes sense.
